src/main/kotlin/com/terraformation/backend/tracking/db/PlantingSiteStore.kt

```diff
@@ -979,4 +986,77 @@ class PlantingSiteStore(
           }
     }
   }
+
+  private fun startPlantingSeason(
+      plantingSiteId: PlantingSiteId,
+      plantingSeasonId: PlantingSeasonId
+  ) {
+    dslContext.transaction { _ ->
+      val rowsUpdated =
+          dslContext
+              .update(PLANTING_SEASONS)
+              .set(PLANTING_SEASONS.IS_ACTIVE, true)
+              .where(PLANTING_SEASONS.ID.eq(plantingSeasonId))
+              .and(PLANTING_SEASONS.PLANTING_SITE_ID.eq(plantingSiteId))
+              .and(PLANTING_SEASONS.IS_ACTIVE.isFalse)
+              .execute()
+
+      if (rowsUpdated > 0) {
+        log.info("Planting season $plantingSeasonId at site $plantingSiteId has started")
+
+        eventPublisher.publishEvent(PlantingSeasonStartedEvent(plantingSiteId, plantingSeasonId))
+      }
+    }
+  }
+
+  private fun startPlantingSeasons() {
+    val now = clock.instant()
+
+    dslContext
+        .select(
+            PLANTING_SEASONS.PLANTING_SITE_ID.asNonNullable(), PLANTING_SEASONS.ID.asNonNullable())
+        .from(PLANTING_SEASONS)
+        .where(PLANTING_SEASONS.START_TIME.le(now))
+        .and(PLANTING_SEASONS.END_TIME.gt(now))
+        .and(PLANTING_SEASONS.IS_ACTIVE.isFalse)
+        .fetch()
+        .forEach { (plantingSiteId, plantingSeasonId) ->
+          startPlantingSeason(plantingSiteId, plantingSeasonId)
+        }
+  }
+
+  private fun endPlantingSeason(
+      plantingSiteId: PlantingSiteId,
+      plantingSeasonId: PlantingSeasonId
+  ) {
+    dslContext.transaction { _ ->
+      val rowsUpdated =
+          dslContext
+              .update(PLANTING_SEASONS)
+              .set(PLANTING_SEASONS.IS_ACTIVE, false)
+              .where(PLANTING_SEASONS.ID.eq(plantingSeasonId))
+              .and(PLANTING_SEASONS.PLANTING_SITE_ID.eq(plantingSiteId))
+              .and(PLANTING_SEASONS.IS_ACTIVE.isTrue)
+              .execute()
+
+      if (rowsUpdated > 0) {
+        log.info("Planting season $plantingSeasonId at site $plantingSiteId has ended")
+
+        eventPublisher.publishEvent(PlantingSeasonEndedEvent(plantingSiteId, plantingSeasonId))
+      }
+    }
+  }
+
+  private fun endPlantingSeasons() {
+    dslContext
+        .select(
+            PLANTING_SEASONS.PLANTING_SITE_ID.asNonNullable(), PLANTING_SEASONS.ID.asNonNullable())
+        .from(PLANTING_SEASONS)
+        .where(PLANTING_SEASONS.END_TIME.le(clock.instant()))
+        .and(PLANTING_SEASONS.IS_ACTIVE.isTrue)
+        .fetch()
+        .forEach { (plantingSiteId, plantingSeasonId) ->
+          endPlantingSeason(plantingSiteId, plantingSeasonId)
```

Comment:
Does this produce another database call in a loop, or is there some `dslContext` ORM functionality going on that is combining the queries?

Reply:
We don't want to combine calls here because we want each transition to be in its own database transaction. Otherwise a failure would cause successful transitions from the same run to be rolled back.

Reply:
Ah yes, I am now recalling that you have already given me this exact answer to this same question before. Sorry for the repetition!